vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/DatabaseBackedProcessor.java

```diff
@@ -1008,6 +1024,233 @@ protected final Set<String> recoveryFailures() {
+        DSL.using(connection, SQLDialect.POSTGRES)
+            .transaction(
+                context -> {
+                  DSLContext dsl = DSL.using(context);
+                  Result<Record> result = dsl.select()
+                      .from(
+                          WORKFLOW_RUN
+                              .join(WORKFLOW_VERSION)
+                              .on(WORKFLOW_RUN.WORKFLOW_VERSION_ID.eq(WORKFLOW_VERSION.ID))
+                              .join(WORKFLOW_DEFINITION)
+                              .on(WORKFLOW_VERSION.WORKFLOW_DEFINITION.eq(WORKFLOW_DEFINITION.ID)))
+                      .where(WORKFLOW_RUN.HASH_ID.eq(workflowRunId)
+                          .and(WORKFLOW_RUN.COMPLETED.isNotNull()))
+                      .fetch();
+                  if (result.isNotEmpty() && result.size() == 1) {
+                    Record record = result.get(0);
+                    // Surely there is a cleaner way to do this
+                    // TODO like a million risk of casting errors
+                    JsonNode metadata = record.get(WORKFLOW_RUN.METADATA);
+                    OffsetDateTime originalCompleted = record.get(WORKFLOW_RUN.COMPLETED);
+                    Iterator<Entry<String, JsonNode>> iterator = metadata.fields();
+                    while (iterator.hasNext()) {
+                      Entry<String, JsonNode> entry = iterator.next();
+                      ArrayNode contents = (ArrayNode) entry.getValue().get("contents");
+                      Iterator<JsonNode> iterator2 = contents.elements();
+                      while (iterator2.hasNext()) {
+                        ObjectNode content = (ObjectNode) iterator2.next();
+                        if (content.has("outputDirectory")) {
+                          content.set("originalDirectory", content.get("outputDirectory"));
+                          content.put("outputDirectory", outputPath);
+                        }
+                      }
+                    }
+                    dsl.update(WORKFLOW_RUN)
+                        .set(WORKFLOW_RUN.COMPLETED,
+                            (OffsetDateTime) null) // Have to cast null to something to resolve ambiguous call lol
+                        .set(WORKFLOW_RUN.METADATA, metadata)
+                        .where(WORKFLOW_RUN.HASH_ID.eq(record.get(WORKFLOW_RUN.HASH_ID)))
+                        .execute();
+
+                    Optional<WorkflowInformation> definition;
+                    try {
+                      definition = getWorkflowByName(
+                          record.get(WORKFLOW_VERSION.NAME),
+                          record.get(WORKFLOW_VERSION.VERSION), dsl);
+                    } catch (SQLException e) {
+                      throw new RuntimeException(e);
+                    }
+
+                    // Get the analysis, and also the external ids
+                    // TODO this could probably be so much more optimized
```

Comment:
The method I can think of for reducing the number of database queries from this is to instead do a monster SELECT with array_agg up at 1069 which gets all the fields and you then iterate over them - like in `Main.createAnalysisRecords` , without the `literalJsonEntry` wrappers.